devito/yask/transformer.py

```diff
@@ -0,0 +1,178 @@
+from devito.dimension import LoweredDimension
+from devito.ir.iet import FindNodes, Expression
+from devito.ir.support import Backward, Forward
+from devito.logger import yask_warning as warning
+from devito.symbolics import split_affine
+
+from devito.yask import nfac
+
+__all__ = ['yaskizer']
+
+
+def yaskizer(trees, yc_soln):
+    """
+    Populate a YASK compiler solution with the :class:`Expression`s found in an IET.
+
+    The necessary YASK grids are instantiated.
+
+    :param trees: A sequence of offloadable :class:`IterationTree`s, in which the
+                  Expressions are searched.
+    :param yc_soln: The YASK compiler solution to be populated.
+    """
+    # Track all created YASK grids
+    mapper = {}
+
+    # It's up to Devito to organize the equations into a flow graph
+    yc_soln.set_dependency_checker_enabled(False)
+
+    processed = []
+    for tree in trees:
+        # All expressions within `tree`
+        expressions = [i.expr for i in FindNodes(Expression).visit(tree.root)]
+
+        # Attach conditional expression for sub-domains
+        conditions = [(i, []) for i in expressions]
+        for i in tree:
+            if not i.dim.is_Sub:
+                continue
+
+            ydim = nfac.new_domain_index(i.dim.parent.name)
+
+            try:
+                # Are the extremes statically known ?
+                lower = nfac.new_const_number_node(int(i.dim.symbolic_ofs_lower))
+                upper = nfac.new_const_number_node(int(i.dim.symbolic_ofs_upper))
+
+                # Yes! Can go on adding the conditions
+                expr = nfac.new_add_node(nfac.new_first_domain_index(ydim), lower)
+                for _, v in conditions:
+                    v.append(nfac.new_not_less_than_node(ydim, expr))
+                expr = nfac.new_add_node(nfac.new_last_domain_index(ydim), upper)
+                for _, v in conditions:
+                    v.append(nfac.new_not_greater_than_node(ydim, expr))
+            except TypeError:
+                # Nope, is at least the Iteration extent statically known? If not,
+                # than a further TypeError exception will be thrown by ``i.extent()``
+                if i.is_Sequential:
+                    extent = i.extent()
+
+                    if i.direction is Backward:
+                        _range = reversed(range(extent))
+                        node = nfac.new_first_domain_index(ydim)
+                    elif i.direction is Forward:
+                        _range = range(-extent + 1, 1)
+                        node = nfac.new_last_domain_index(ydim)
+                    else:
```

Comment:
Is this assuming that Forward means "right boundary" and Backward means "left boundary"?  That works for us, but I'm not sure it's a safe assumption, in general.

Reply:
See `test_yask.py::TestOperatorSimple::test_subdimleft_notparallel` for a contrived example.